### PR TITLE
Parameter derivatives for SOECPotential

### LINKED ISSUE #4297: Jastrow optimization with SOECP no longer working
**Describe the bug**
Since QMCPACK 3.15.0, Jastrow optimization with SOECP is giving failure with following message,

terminate called after throwing an instance of 'std::logic_error'
  what():  Bug!! SOECPotential::evaluateValueAndDerivativesmust be overloaded when the OperatorBase depends on a wavefunction.

As it normally works on QMCPACK 3.14.0, a trigger looks to be located between 3.14.0 and 3.15.0.

**To Reproduce**
Try VTe2 optimization located on
https://anl.box.com/s/okkc0r0iuvhe5fzbd63kte62jdwqk8gr

**Expected behavior**
Jastrow optimization should work normally with SOECP.

**System:**
 - system name 
 theta, summit

### PR DESCRIPTION
**Pull request by camelto2**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

This PR implements parameter derivatives for the SOECPotential. This closes #4297 

Note that I had to modify the SOECP unit test a bit in order to have a wave function that could be optimized. The new reference values are all obtained from the standalone code soecp_val_reference.cpp in QMCHamiltonians/tests. This implements everything from scratch and uses finite differences to obtain the reference values. 

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
M1 mac

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
Test this please

File: src/QMCHamiltonians/SOECPComponent.cpp [hunk 1]
```diff
@@ -193,6 +190,93 @@ SOECPComponent::RealType SOECPComponent::evaluateOne(ParticleSet& W,
...
+                                                                     const PosType& dr,
+                                                                     const opt_variables_type& optvars,
+                                                                     const Vector<ValueType>& dlogpsi,
+                                                                     Vector<ValueType>& dhpsioverpsi)
+{
+#ifndef QMC_COMPLEX
+  throw std::runtime_error("SOECPComponent::evaluateValueAndDerivatives should not be called in real build\n");
+#else
+  if (sknot_ < 2)
+    APP_ABORT("Spin knots must be greater than 2\n");
```
Comment (ye-luo) [inline src/QMCHamiltonians/SOECPComponent.cpp:207]: Not "greater equal"?
Could you move this check where sknot_ is set?
APP_ABORT better using exception.

**Comment by camelto2** [comment]
Tests were failing with the real build for my initial PR because I have data that needs to interface with WF APIs and thus need to be ValueType....but part of the SOECP eval is inherently complex. Problem line was

wvec_[iq] = lsum * psiratio_[iq] * spin_quad_weights_[iq];

lsum has to be ComplexType (due to the ldots matrix element that is carried out) but wvec_ needs to be ValueType for the subsequent gemv call to get the dhpsioverpsi correct. 

I wasn't sure of the cleanest way to fix this...basically none of the spinor and SOECP stuff should ever be used if it is a real build in the first place but it seems we don't have perfect guards in place. I tried to modify the CMakeLists to only build SOECPotential/SOECPComponent if QMC_COMPLEX, but that left a lot undefined symbols due to the ECPotential/ECPComponent builders. So I resorted to adding an ifdef for now in the evaluateValueAndDerivatives...which I don't like because I know we are trying to move away from that but that was the simplest solution I came up with

File: src/QMCHamiltonians/SOECPComponent.cpp (hunk 1 above)
Comment (camelto2) [reply src/QMCHamiltonians/SOECPComponent.cpp]: fixed

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**